### PR TITLE
Enhance circle ci tests

### PR DESCRIPTION
**Pull request by NicolasDP**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -8,71 +8,106 @@ jobs:
       - checkout
       - run: git submodule sync
       - run: git submodule update --init
-      - restore_cache:
```
Comment (mzabaluev) [inline .circleci/config.yml]: Not restoring cache from past builds, even if those don't completely match this build's `Cargo.lock`, will make every build job re-fetch every crate.

File: .circleci/config.yml [hunk 2]
```diff
@@ -8,71 +8,106 @@ jobs:
...
-          keys:
-            - cargo-v2-{{ checksum "Cargo.toml" }}-
-            - cargo-v2-
       - run: cargo fetch
-      - persist_to_workspace:
-          root: "."
-          paths:
-            - Cargo.lock
```
Comment (mzabaluev) [inline .circleci/config.yml]: OK since `Cargo.lock` is version-controlled.

File: .circleci/config.yml [hunk 3]
```diff
@@ -8,71 +8,106 @@ jobs:
...
+      - run:
+          name: Print version information
+          command: rustc --version; cargo --version
+      - run:
+          name: Build
+          command: cargo build --frozen
+      - persist_to_workspace:
+          root: target
+          paths:
+            - debug/*
```
Comment (mzabaluev) [inline .circleci/config.yml]: Consider pruning it to only persisting the cached sources; the test will rebuild the binaries anyway.

File: .circleci/config.yml [hunk 4]
```diff
@@ -8,71 +8,106 @@ jobs:
...
-  test:
+  rustfmt:
+    docker:
+      - image: rust:latest
+    working_directory: /mnt/crate
+    steps:
+      - checkout
+      - run:
+          name: Install rustfmt
+          command: rustup component add rustfmt
```
Comment (mzabaluev) [inline .circleci/config.yml]: Ideally, we'd want a separate container image with rustfmt pre-installed to save build time. This will do for now.

File: .circleci/config.yml (hunk 1 above)
Comment (NicolasDP) [reply .circleci/config.yml]: thanks, I've put back the restore cache here too now

File: .circleci/config.yml [hunk 5]
```diff
@@ -6,73 +6,125 @@ jobs:
     working_directory: /mnt/crate
     steps:
       - checkout
-      - run: git submodule sync
-      - run: git submodule update --init
       - restore_cache:
           keys:
-            - cargo-v2-{{ checksum "Cargo.toml" }}-
-            - cargo-v2-
+            - cargo-cache-{{ checksum "Cargo.toml" }}-{{ checksum "Cargo.lock" }}
```
Comment (mzabaluev) [inline .circleci/config.yml]: Good idea to add this one.

File: .circleci/config.yml [hunk 6]
```diff
@@ -6,73 +6,125 @@ jobs:
...
+            - cargo-cache-{{ checksum "Cargo.toml" }}
+            - cargo-cache-
```
Comment (mzabaluev) [inline .circleci/config.yml]: Why did you rename the prefix? The old `cargo-v2-*` entries are still valid for this cache, and the versioning is a useful practice for forcing a fresh cache when we decide we don't need some previously cached directories.

Comment (NicolasDP) [reply .circleci/config.yml]: re-renamed

File: .circleci/config.yml (hunk 3 above)
Comment (mzabaluev) [reply .circleci/config.yml]: The binaries might be reused as well, at least accordingly to my quick comparison of the build directory between the steps.
